src/Interpreters/ProcessList.cpp

```diff
@@ -541,6 +558,28 @@ CancellationCode ProcessList::sendCancelToQuery(const String & current_query_id,
 }
 
 
+CancellationCode ProcessList::sendCancelToQuery(QueryStatusPtr elem, bool kill)
+{
+    /// Cancelling the query should be done without the lock.
+    /// So here we first set is_cancelling, and later reset it.
+    /// The ProcessListEntry cannot be destroy if is_cancelling is true.
+    {
+        auto lock = safeLock();
+        elem->is_cancelling = true;
+    }
+
+    SCOPE_EXIT({
+        DENY_ALLOCATIONS_IN_SCOPE;
+
+        auto lock = unsafeLock();
+        elem->is_cancelling = false;
+        cancelled_cv.notify_all();
+    });
+
```

Comment:
It looks more logically consistent now - because first we set `is_cancelling` to true, then we call `cancelQuery(kill)` and set `is_cancelling` to false. Anyway the difference is not quite essential because `safeLock()` should never throw.